### ISSUE TITLE
Types of TauSampling/MatsubaraSampling

### ISSUE THREAD

**Issue by shinaoka** [SpM-lab/SparseIR.jl#28]
These types depend on the SVD type:
https://github.com/SpM-lab/SparseIR.jl/blob/main/src/sampling.jl#L69

This is not so convenient because changes in the implementation of the fitting could affect user codes. For instance, the code shown below does not work anymore.
https://github.com/SpM-lab/sparse-ir-tutorial/blob/main/src/ipt_jl.md

![image](https://user-images.githubusercontent.com/15949860/171524801-0fb1384b-383e-4230-9e06-635ebff0d851.png)

Does it make sense to define type aliases?

```
const TauSampling64 = TauSampling{Float64,Float64,SVD}
const MatsubaraSampling64 = MatsubaraSampling{Int64,ComplexF64,SVD}
```

**Comment by SamuelBadr** [comment]
A more idiomatic way of doing this might be the following (untested):
```julia
struct IPTSolver{T<:AbstractFloat,FTB<:FiniteTempBasis,F<:Function,MS<:MatsubaraSampling,TS<:TauSampling}
    U::T
    basis::FTB
    beta::T
    rho_omega::F
    omega_range::Tuple{T,T}
    quad_rule::Tuple{Vector{T},Vector{T}}
    smpl_matsu::MS
    smpl_tau::TS
end

function IPTSolver(U::T, basis, rho, omega_range::Tuple{T,T}, deg_leggaus=100) where {T<:AbstractFloat}
    quad_rule = _gausslegendre(deg_leggaus, omega_range...)
    smpl_matsu = MatsubaraSampling(basis)
    smpl_tau = TauSampling(basis)

    IPTSolver(U, basis, SparseIR.getbeta(basis), rho_omega, omega_range, quad_rule, smpl_matsu, smpl_tau)
end
```
(Note that `rho_omega` will now be concretely typed whereas it was abstract before, leading to potentially poor performance.)
Note also that the parameter constraints are not strictly needed -- we could also have written `IPTSolver{T,FTB,F,MS,TS}` -- and should be specified only when it helps convey intent, as I think it does here. 
This kind of parametrized code avoids having to wrestle with more complicated library types' parameters and will be more easily adaptable to changes.

Because `SVD` is not a concrete type itself, your proposal would need to be extended to `const TauSampling64 = TauSampling{Float64, Float64, SVD{Float64, Float64, Matrix{Float64}, Vector{Float64}}}` in order to be a concrete type, but in general I'd prefer keeping things parametric and avoid specific types (like `Float64`) as much as possible to make the code useable with different inputs, like `Float64x2` if the user deems it preferable.

BTW, quickly looking over the linked notebook I noticed that every function argument is completely specified. This is not necessary in Julia (https://docs.julialang.org/en/v1/manual/performance-tips/#Type-declarations).

**Comment by shinaoka** [comment]
I got your point. I'd also prefer to keep things parametric in the library. Nevertheless, in a user's program (not a library), it's not necessary for them to know what is the backend of the fitting.

Now, we both are OK to introduce such type aliases?
```
const TauSampling64 = TauSampling{Float64, Float64, SVD{Float64, Float64, Matrix{Float64}, Vector{Float64}}}
```

> BTW, quickly looking over the linked notebook I noticed that every function argument is completely specified. This is not necessary in Julia 

I know, but some people from C++/Fortran prefer this style... In my opinion, one good of Julia is that it's easy to use different programming styles. I also prefer to specify function arguments outside the library.

**Comment by shinaoka** [comment]
BTW, I have recently found an interesting library: 
https://github.com/jonniedie/ConcreteStructs.jl

**Comment by SamuelBadr** [comment]
> Now, we both are OK to introduce such type aliases?
> 
> ```
> const TauSampling64 = TauSampling{Float64, Float64, SVD{Float64, Float64, Matrix{Float64}, Vector{Float64}}}
> ```
Yep, for sure! 

> BTW, I have recently found an interesting library:
> https://github.com/jonniedie/ConcreteStructs.jl

I'll take a look at that, thank you!

**Comment by SamuelBadr** [comment]
> BTW, I have recently found an interesting library:
> https://github.com/jonniedie/ConcreteStructs.jl

This seems really nice, I'll definitely be using it! Wonder why something like it is not in the language itself...

**Comment by shinaoka** [comment]
Addressed in e8763e3088688d50039a6673d1207703c4c6fec5
